File: trunk/tests/Pdbg/ObservableTest.php
<?php
/**
 * PDbg, A DBGp debugger client
 * Copyright (C) 2008, Christopher Utz <cutz@chrisutz.com>
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 * 
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 * 
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 *
 * @category   Development
 * @package    Pdbg
 * @author     Christopher Utz <cutz@chrisutz.com>
 * @copyright  2008 Christopher Utz <cutz@chrisutz.com>
 * @license    http://www.gnu.org/licenses/gpl.html GPLv3
 * @version    SVN: $Id$
 * @link       http://pdbg.googlecode.com
 */

/**
 * @see Pdbg_Observable
 */
require_once 'Pdbg/Observable.php';

/**
 * Tests for Pdbg_Observable
 *
 * @category   Development
 * @package    Pdbg
 * @author     Christopher Utz <cutz@chrisutz.com>
 * @copyright  2008 Christopher Utz <cutz@chrisutz.com>
 * @license    http://www.gnu.org/licenses/gpl.html GPLv3
 * @version    SVN: $Id$
 * @link       http://pdbg.googlecode.com
 */
class Pdbg_ObservableTest extends PHPUnit_Framework_TestCase
{
    /**
     * Test to ensure that addEvent allows method chaining.
     *
     * @return void
     */
    public function testAddEvent()
    {
        $observable = new Pdbg_Observable();
        $this->assertType('Pdbg_Observable', $observable->addEvent('foo'));
    }

    /**
     * Test to ensure that addObserver allows method chaining.
     *
     * @return void
     */
    public function testAddObserver()
    {
        $observable = new Pdbg_Observable();
        $observable->addEvent('foo');

        $fn = create_function('', '');

        $this->assertType('Pdbg_Observable', $observable->addObserver('foo', $fn));
    }

    /**
     * Test to ensure that firing events works as expected with a single
     * argument.
     *
     * @return void
     */
    public function testFireOneArg()
    {
        $observable = new Pdbg_Observable();

        $fn = create_function('$a', 'echo $a;');
        
        $observable->addEvent('foo')
            ->addEvent('bar')
            ->addObserver('foo', $fn)
            ->addObserver('foo', $fn)
            ->addObserver('bar', $fn);

        ob_start();
        $observable->fire('foo', '-');
        $result = ob_get_clean();

        $this->assertEquals('--', $result);
    }

    /**
     * Test to ensure that firing events works as expected with multiple
     * arguments.
     *
     * @return void
     */
    public function testFireArgs()
    {
        $observable = new Pdbg_Observable();

        $fn = create_function('$a, $b', 'echo $a . $b;');
        
        $observable->addEvent('foo')->addObserver('foo', $fn);

        ob_start();
        $observable->fire('foo', array('-', '>'));
        $result = ob_get_clean();

        $this->assertEquals('->', $result);
    }
}

File: branches/python-rewrite/pdbg/observabletest.py
# Written by Christopher Utz <cutz@chrisutz.com>
# See LICENSE.txt for license information

"""Unit tests for the observable module."""

__version__ = "$Id$"

import unittest
from observable import *

def _new_inc_observer():
    def f(observable, amt=1):
        f.times += amt
    f.times = 0
    return f

class TestObservableClass(unittest.TestCase):
    """Test the Observable class."""

    def test_fire(self):
        f = _new_inc_observer()
        o = Observable()
        o.register_event('foo')
        o.add_observer('foo', f)
        self.assertEqual(f.times, 0)
        o.fire('foo')
        self.assertEqual(f.times, 1)
        o.fire('foo')
        self.assertEqual(f.times, 2)

    def test_two_observers(self):
        f = _new_inc_observer()
        o = Observable()
        o.register_event('foo')
        o.add_observer('foo', f).add_observer('foo', f)
        o.fire('foo')
        self.assertEqual(f.times, 2)

    def test_two_events(self):
        f = _new_inc_observer()
        o = Observable()
        o.register_event('foo', 'bar')
        o.add_observer(foo=f, bar=f)
        o.fire('foo').fire('bar')
        self.assertEqual(f.times, 2)

    def test_fire_param(self):
        f = _new_inc_observer()
        o = Observable()
        o.register_event('foo', 'bar')
        o.add_observer(foo=f, bar=f)
        o.fire('foo').fire('bar', 2)
        self.assertEqual(f.times, 3)

    def test_remove_observer(self):
        f1 = _new_inc_observer()
        f2 = _new_inc_observer()
        o = Observable()
        o.register_event('foo')
        o.add_observer('foo', f1).add_observer('foo', f2)
        self.assertEqual(o.remove_observer('foo', f2), True)
        o.fire('foo')
        self.assertEqual(f1.times, 1)
        self.assertEqual(f2.times, 0)

if __name__ == '__main__':
    unittest.main()
